### ISSUE TITLE
A few issues on OS X 10.11 El Capitan

### ISSUE THREAD

**Issue by vollkommenheit** [DNS-OARC/dnscap#1]
Hi, 

I'm not a programmer, just an enthusiast and github newbie. I thought I have patched around two issues in trying to compile dnscap on my OS X 10.11 El Capitan system. However, in doing so I may have introduced a new one. Please see below.

Issue 1: I encountered it during "make" with both https://dnscap.dns-oarc.net/dnscap-20150206.zip and the https://github.com/verisign/dnscap fork:
> gcc -Wall -g -O2  -c dump_dns.c
dump_dns.c:63:9: warning: 'p_rcode' macro redefined [-Wmacro-redefined]
#define p_rcode __p_rcode
        ^
/usr/include/resolv.h:386:9: note: previous definition is here
#define p_rcode                                 res_9_p_rcode
        ^
dump_dns.c:189:4: error: use of undeclared identifier 'INT32SZ'
                        MY_GET32(soa[n], rd);
                        ^
dump_dns.c:85:10: note: expanded from macro 'MY_GET32'
        (cp) += INT32SZ; \
                ^
dump_dns.c:200:3: error: use of undeclared identifier 'INT16SZ'
                MY_GET16(mx, rd);
                ^
dump_dns.c:75:10: note: expanded from macro 'MY_GET16'
        (cp) += INT16SZ; \
                ^
1 warning and 2 errors generated.
make: *** [dump_dns.o] Error 1

I think I've patched around this, [dump_dns.c.diff.txt](https://github.com/DNS-OARC/dnscap/files/262286/dump_dns.c.diff.txt), using bits from https://github.com/gamelinux/prads/blob/master/src/dump_dns.c

Issue 2: I only encountered it with the https://github.com/verisign/dnscap fork, apparently due to the most recent https://github.com/verisign/dnscap/commit/555b73cf6c8b9111c251c5d20fc5ed78f8e62049 commit, which adds syscalls that are named differently on OS X:
> gcc -Wall -I/usr/local/include  -c dnscap.c
dnscap.c:488:6: warning: implicit declaration of function 'setresgid' is invalid in C99 [-Wimplicit-function-declaration]
        if (setresgid(dropGID, dropGID, dropGID) < 0) {
            ^
dnscap.c:493:6: warning: implicit declaration of function 'setresuid' is invalid in C99 [-Wimplicit-function-declaration]
        if (setresuid(dropUID, dropUID, dropUID) < 0) {
            ^
dnscap.c:2485:6: warning: explicitly assigning value of variable of type 'int' to itself [-Wself-assign]
                        x = x;
                        ~ ^ ~
dnscap.c:8:19: warning: unused variable 'copyright' [-Wunused-const-variable]
static const char copyright[] =
                  ^
4 warnings generated.
gcc -Wall -I/usr/local/include  -c dump_dns.c
dump_dns.c:12:19: warning: unused variable 'rcsid' [-Wunused-const-variable]
static const char rcsid[] = "$Id: dump_dns.c,v 1.2 2008-03-14 21:33:28 wessels Exp $";
                  ^
1 warning generated.
gcc -o dnscap -L/usr/local/lib dnscap.o dump_dns.o  -lresolv -ldl -lpcap
Undefined symbols for architecture x86_64:
  "_setresgid", referenced from:
      _drop_privileges in dnscap.o
  "_setresuid", referenced from:
      _drop_privileges in dnscap.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [dnscap] Error 1

I thought I have patched around it, [dnscap.c.diff.txt](https://github.com/DNS-OARC/dnscap/files/262295/dnscap.c.diff.txt), by adapting bits of http://ab-initio.mit.edu/octave-Faddeeva/gnulib/lib/idpriv-drop.c. However, in doing so, I have introduced a new issue into the resulting dnscap binary.
> $ sudo dnscap -d -I -g -6 -i en0
Password:
dnscap: version V1.0-OARC-r%d (%s)
dnscap: msg Q.., side IR, hide .., err NYtfsxir, t 0, c 0
dnscap: "( ( ip proto 1 or ip proto 58 ) or ( ( ( udp port 53) ) ) )"
Unable to drop GID to nobody, exiting.

Upon a closer look, the patch [dnscap.c.diff.txt](https://github.com/DNS-OARC/dnscap/files/262295/dnscap.c.diff.txt) is not valid. Need a real fix for the below:

> gcc -Wall -g -O2  -c dnscap.c
dnscap.c: In function 'drop_privileges':
dnscap.c:448:8: warning: variable 'dropGID' set but not used [-Wunused-but-set-variable]
  gid_t dropGID;
        ^
dnscap.c:447:8: warning: variable 'dropUID' set but not used [-Wunused-but-set-variable]
  uid_t dropUID;
        ^
gcc -Wall -g -O2  -c dump_dns.c
dump_dns.c:70:0: warning: "p_rcode" redefined
 #define p_rcode __p_rcode
 ^
In file included from dump_dns.c:62:0:
/usr/include/resolv.h:386:0: note: this is the location of the previous definition
 #define p_rcode     res_9_p_rcode
 ^
gcc -o dnscap  dnscap.o dump_dns.o  -ldl -lpcap -lresolv -lbind
nroff -mandoc dnscap.1 > dnscap.cat1

**Comment by vollkommenheit** [comment]
It seems replacing the block below in dnscap.c with the second block should let dropGID/dropUID succeed for non-setresgid/setresuid platforms such as OS X, with the caveat that ./configure needs to create a "config.h" that tests for HAVE_SETRESGID, HAVE_SETREGID, HAVE_SETEGID, etc, which I haven't figured out how to do yet.

``
	if (setresgid(dropGID, dropGID, dropGID) < 0) {
		fprintf(stderr, "Unable to drop GID to %s, exiting.\n", DROPTOUSER);
		exit(1);
	}

	if (setresuid(dropUID, dropUID, dropUID) < 0) {
		fprintf(stderr, "Unable to drop UID to %s, exiting.\n", DROPTOUSER);
		exit(1);
	}

`` 	/* Drop the gid privilege first, because in some cases the gid privilege
	   cannot be dropped after the uid privilege has been dropped.  */

	/* This is for executables that have the setgid bit set.  */
	#if HAVE_SETRESGID /* glibc, FreeBSD, OpenBSD, HP-UX */
	/* This code is needed: In particular, on HP-UX 11.11, setregid (gid, gid)
	   may leave the saved gid as 0.  See also the comment below regarding
	   setresuid.  */
	if (setresgid(dropGID, dropGID, dropGID) < 0)
	{
 		fprintf(stderr, "Unable to drop GID to %s, exiting.\n", DROPTOUSER);
 		exit(1);
 	}
	#elif HAVE_SETREGID /* Mac OS X, NetBSD, AIX, IRIX, Solaris, OSF/1, Cygwin */
	if (setregid (dropGID, dropGID) < 0)
	{
 		fprintf(stderr, "Unable to drop GID to %s, exiting.\n", DROPTOUSER);
 		exit(1);
 	}
	#elif HAVE_SETEGID /* Solaris 2.4 */
	if (setegid (dropGID) < 0)
	{
 		fprintf(stderr, "Unable to drop GID to %s, exiting.\n", DROPTOUSER);
 		exit(1);
 	}
	#endif

	/* This is for executables that have the setuid bit set.  */
	#if HAVE_SETRESUID /* glibc, FreeBSD, OpenBSD, HP-UX */
	/* On systems which have setresuid(), we use it instead of setreuid(),
	   because
	      <http://www.usenix.org/events/sec02/full_papers/chen/chen.pdf>
	   says about setreuid(): "The rule by which the saved uid id is modified
	   is complicated." Similarly, <http://unixpapa.com/incnote/setuid.html>
	   says about setreuid(): "What exactly happens to the saved UID when this
	   is used seems to vary a lot."  */
	if (setresuid(dropUID, dropUID, dropUID) < 0)
	{
 		fprintf(stderr, "Unable to drop UID to %s, exiting.\n", DROPTOUSER);
 		exit(1);
 	}
	#elif HAVE_SETREUID /* Mac OS X, NetBSD, AIX, IRIX, Solaris, OSF/1, Cygwin */
	if (setreuid (dropUID, dropUID) < 0)
	{
 		fprintf(stderr, "Unable to drop UID to %s, exiting.\n", DROPTOUSER);
 		exit(1);
 	}
	#elif HAVE_SETEUID /* Solaris 2.4 */
	if (seteuid (dropUID) < 0)
	{
 		fprintf(stderr, "Unable to drop UID to %s, exiting.\n", DROPTOUSER);
 		exit(1);
 	}
	#endif

**Comment by vollkommenheit** [comment]
Got it. Just need to add the following line to configure.in, followed by "autoconf" to create a new "configure" script. 

`AC_CHECK_FUNCS([setresgid setresuid setregid setreuid setegid seteuid ])
`

Now all the "#if HAVE_SETRESGID" checks are recognized, and the resulting dnscap binary drops to nobody successfully on OS X 10.11:

> nobody          12369   0.0  0.0  2463604    884 s006  S+    8:44PM   0:00.00 dnscap -d -I -g -6 -i en1
root            12368   0.0  0.0  2471648   4288 s006  U+    8:44PM   0:00.02 sudo dnscap -d -I -g -6 -i en1

**Comment by jelu** [comment]
Thanks for the patch, we are in the process to setup management of this repository and I will include this fix once that has been done.

**Comment by jelu** [comment]
This has partly been fixed by #5.

**Comment by vollkommenheit** [comment]
It seems dump_dns.c still needs:

> +#ifndef INT16SZ
+#define INT16SZ (2)
+#endif
+#ifndef INT32SZ
+#define INT32SZ (4)
+#endif

in order to avoid:

> dump_dns.c: In function ‘dump_dns_rr’:
dump_dns.c:90:10: error: ‘INT32SZ’ undeclared (first use in this function)
  (cp) += INT32SZ; \
          ^
dump_dns.c:194:4: note: in expansion of macro ‘MY_GET32’
    MY_GET32(soa[n], rd);
    ^
dump_dns.c:90:10: note: each undeclared identifier is reported only once for each function it appears in
  (cp) += INT32SZ; \
          ^
dump_dns.c:194:4: note: in expansion of macro ‘MY_GET32’
    MY_GET32(soa[n], rd);
    ^
dump_dns.c:80:10: error: ‘INT16SZ’ undeclared (first use in this function)
  (cp) += INT16SZ; \
          ^
dump_dns.c:211:3: note: in expansion of macro ‘MY_GET16’
   MY_GET16(mx, rd);
   ^
make[2]: *** [dump_dns.o] Error 1
make[1]: *** [all] Error 2
make: *** [all-recursive] Error 1

**Comment by jelu** [comment]
Thanks for testing, please test develop branch again.

**Comment by vollkommenheit** [comment]
All good now. There're a few warnings, which I assume are negligible:

> Making all in src
/Applications/Xcode.app/Contents/Developer/usr/bin/make  all-am
depbase=`echo dnscap.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
	gcc -DHAVE_CONFIG_H -I.    -I. -I..  -g -O2 -MT dnscap.o -MD -MP -MF $depbase.Tpo -c -o dnscap.o dnscap.c &&\
	mv -f $depbase.Tpo $depbase.Po
depbase=`echo dump_dns.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
	gcc -DHAVE_CONFIG_H -I.    -I. -I..  -g -O2 -MT dump_dns.o -MD -MP -MF $depbase.Tpo -c -o dump_dns.o dump_dns.c &&\
	mv -f $depbase.Tpo $depbase.Po
sed -e 's,[@]PACKAGE_VERSION[@],2.0.0,g' \
        -e 's,[@]PACKAGE_URL[@],https://github.com/DNS-OARC/dnscap/issues,g' \
        -e 's,[@]PACKAGE_BUGREPORT[@],dnscap-users@dns-oarc.net,g' \
        < ./dnscap.1.in > dnscap.1
/bin/sh ../libtool  --tag=CC   --mode=link gcc -I. -I..  -g -O2   -o dnscap dnscap.o dump_dns.o   -ldl -lpcap -lresolv -lbind
libtool: link: gcc -I. -I.. -g -O2 -o dnscap dnscap.o dump_dns.o  -ldl -lpcap -lresolv -lbind
ld: warning: could not create compact unwind for ___ns_sprintrrf: stack subq instruction is too different from dwarf stack size
ld: warning: could not create compact unwind for ___res_nmkupdate: stack subq instruction is too different from dwarf stack size
ld: warning: could not create compact unwind for ___res_nupdate: stack subq instruction is too different from dwarf stack size
ld: warning: could not create compact unwind for _dst_s_write_public_key: stack subq instruction is too different from dwarf stack size
ld: warning: could not create compact unwind for ___ns_verify: stack subq instruction is too different from dwarf stack size
ld: warning: could not create compact unwind for ___ns_verify_tcp: stack subq instruction is too different from dwarf stack size
ld: warning: could not create compact unwind for _save_ns: stack subq instruction is too different from dwarf stack size
ld: warning: could not create compact unwind for ___res_findzonecut2: stack subq instruction is too different from dwarf stack size
ld: warning: could not create compact unwind for _get_txt_records.isra.0: stack subq instruction is too different from dwarf stack size
Making all in plugins
Making all in pcapdump
depbase=`echo pcapdump.lo | sed 's|[^/]*$|.deps/&|;s|\.lo$||'`;\
	/bin/sh ../../libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I../../src    -I. -I../../src -I../../isc  -g -O2 -MT pcapdump.lo -MD -MP -MF $depbase.Tpo -c -o pcapdump.lo pcapdump.c &&\
	mv -f $depbase.Tpo $depbase.Plo
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I../../src -I. -I../../src -I../../isc -g -O2 -MT pcapdump.lo -MD -MP -MF .deps/pcapdump.Tpo -c pcapdump.c  -fno-common -DPIC -o .libs/pcapdump.o
/bin/sh ../../libtool  --tag=CC   --mode=link gcc -I. -I../../src -I../../isc  -g -O2 -module -version-info 2:0:0  -o pcapdump.la -rpath /usr/local/lib/dnscap pcapdump.lo  -ldl -lpcap -lresolv -lbind
libtool: link: gcc -Wl,-undefined -Wl,dynamic_lookup -o .libs/pcapdump.2.so -bundle  .libs/pcapdump.o   -ldl -lpcap -lresolv -lbind  -g -O2
libtool: link: (cd ".libs" && rm -f "pcapdump.so" && ln -s "pcapdump.2.so" "pcapdump.so")
libtool: link: ( cd ".libs" && rm -f "pcapdump.la" && ln -s "../pcapdump.la" "pcapdump.la" )
Making all in rssm
depbase=`echo rssm.lo | sed 's|[^/]*$|.deps/&|;s|\.lo$||'`;\
	/bin/sh ../../libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I../../src    -I. -I../../src -I../../isc  -g -O2 -MT rssm.lo -MD -MP -MF $depbase.Tpo -c -o rssm.lo rssm.c &&\
	mv -f $depbase.Tpo $depbase.Plo
depbase=`echo hashtbl.lo | sed 's|[^/]*$|.deps/&|;s|\.lo$||'`;\
	/bin/sh ../../libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I../../src    -I. -I../../src -I../../isc  -g -O2 -MT hashtbl.lo -MD -MP -MF $depbase.Tpo -c -o hashtbl.lo hashtbl.c &&\
	mv -f $depbase.Tpo $depbase.Plo
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I../../src -I. -I../../src -I../../isc -g -O2 -MT rssm.lo -MD -MP -MF .deps/rssm.Tpo -c rssm.c  -fno-common -DPIC -o .libs/rssm.o
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I../../src -I. -I../../src -I../../isc -g -O2 -MT hashtbl.lo -MD -MP -MF .deps/hashtbl.Tpo -c hashtbl.c  -fno-common -DPIC -o .libs/hashtbl.o
/bin/sh ../../libtool  --tag=CC   --mode=link gcc -I. -I../../src -I../../isc  -g -O2 -module -version-info 2:0:0  -o rssm.la -rpath /usr/local/lib/dnscap rssm.lo hashtbl.lo  -lldns -ldl -lpcap -lresolv -lbind
libtool: link: gcc -Wl,-undefined -Wl,dynamic_lookup -o .libs/rssm.2.so -bundle  .libs/rssm.o .libs/hashtbl.o   /usr/local/lib/libldns.dylib -L/usr/local/lib -lcrypto -ldl -lpcap -lresolv -lbind  -g -O2
libtool: link: (cd ".libs" && rm -f "rssm.so" && ln -s "rssm.2.so" "rssm.so")
libtool: link: ( cd ".libs" && rm -f "rssm.la" && ln -s "../rssm.la" "rssm.la" )
Making all in txtout
depbase=`echo txtout.lo | sed 's|[^/]*$|.deps/&|;s|\.lo$||'`;\
	/bin/sh ../../libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I../../src    -I. -I../../src -I../../isc  -g -O2 -MT txtout.lo -MD -MP -MF $depbase.Tpo -c -o txtout.lo txtout.c &&\
	mv -f $depbase.Tpo $depbase.Plo
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I../../src -I. -I../../src -I../../isc -g -O2 -MT txtout.lo -MD -MP -MF .deps/txtout.Tpo -c txtout.c  -fno-common -DPIC -o .libs/txtout.o
txtout.c:129:41: warning: format specifies type 'unsigned long' but the argument has type '__darwin_suseconds_t' (aka 'int') [-Wformat]
        fprintf(out, "%10lu.%06lu", ts.tv_sec, ts.tv_usec);
                            ~~~~~              ^~~~~~~~~~
                            %06d
1 warning generated.
/bin/sh ../../libtool  --tag=CC   --mode=link gcc -I. -I../../src -I../../isc  -g -O2 -module -version-info 2:0:0  -o txtout.la -rpath /usr/local/lib/dnscap txtout.lo  -ldl -lpcap -lresolv -lbind
libtool: link: gcc -Wl,-undefined -Wl,dynamic_lookup -o .libs/txtout.2.so -bundle  .libs/txtout.o   -ldl -lpcap -lresolv -lbind  -g -O2
libtool: link: (cd ".libs" && rm -f "txtout.so" && ln -s "txtout.2.so" "txtout.so")
libtool: link: ( cd ".libs" && rm -f "txtout.la" && ln -s "../txtout.la" "txtout.la" )
make[2]: Nothing to be done for `all-am'.
make[1]: Nothing to be done for `all-am'.

**Comment by jelu** [comment]
The `txtout` warning should be fixed now but the other stuff is very OS X specific and related to compiler options and not only the options dnscap use but all linked libraries. Luckily it is not a blocking issue, it should work anyway (might be usecs "slower").
